### ISSUE TITLE
Tx-log post request via http module with Kafka JSON producer gives com.fasterxml.jackson.core.JsonGenerationException

### ISSUE THREAD

**Issue by lavsurgut** [xtdb/xtdb#1176]
So sending any document tx via default Nippy SerDe works ok.
E.g. 

```shell
curl -X POST \
     -H "Content-Type: application/edn" \
     -d '[[:crux.tx/put {:crux.db/id :ivan, :name "Ivan" :last-name "Petrov"}],
          [:crux.tx/put {:crux.db/id :boris, :name "Boris" :last-name "Petrov"}],
          [:crux.tx/delete :maria  #inst "2012-05-07T14:57:08.462-00:00"]]' \
     http://crux:3000/tx-log
```

With JSON though:

```
ERROR crux.http-server - Exception while handling request: {:ssl-client-cert nil, :protocol "HTTP/1.1", :remote-addr "10.0.11.28", :headers {"user-agent" "Go-http-client/1.1", "host" "crux:3000", "accept-encoding" "gzip", "content-length" "1131", "content-type" "application/edn"}, :server-port 3000, :content-length 1131, :content-type "application/edn", :character-encoding nil, :uri "/tx-log", :server-name "crux", :query-string nil, :body #object[org.eclipse.jetty.server.HttpInputOverHTTP 0x63b03bf0 "HttpInputOverHTTP@63b03bf0[c=1131,q=0,[0]=null,s=EOF]"], :scheme :http, :request-method :post}
com.fasterxml.jackson.core.JsonGenerationException: Cannot JSON encode object of class: class crux.codec.Id: d827db3da2a4b618261eaba80cdf5f1f584609ee
	at cheshire.generate$generate.invokeStatic(generate.clj:152)
	at cheshire.generate$generate.invoke(generate.clj:116)
	at cheshire.core$generate_string.invokeStatic(core.clj:74)
	at cheshire.core$generate_string.invoke(core.clj:49)
	at cheshire.core$generate_string.invokeStatic(core.clj:55)
	at cheshire.core$generate_string.invoke(core.clj:49)
	at clojure.lang.Var.invoke(Var.java:384)
	at crux.kafka.json.JsonSerializer.serialize(JsonSerializer.java:53)
	at org.apache.kafka.common.serialization.Serializer.serialize(Serializer.java:62)
	at org.apache.kafka.clients.producer.KafkaProducer.doSend(KafkaProducer.java:886)
	at org.apache.kafka.clients.producer.KafkaProducer.send(KafkaProducer.java:856)
	at org.apache.kafka.clients.producer.KafkaProducer.send(KafkaProducer.java:743)
	at crux.kafka$submit_docs.invokeStatic(kafka.clj:231)
	at crux.kafka$submit_docs.invoke(kafka.clj:228)
	at crux.kafka.KafkaDocumentStore.submit_docs(kafka.clj:245)
	at crux.node.CruxNode.submitTx(node.clj:125)
	at crux.http_server$transact.invokeStatic(http_server.clj:195)
	at crux.http_server$transact.invoke(http_server.clj:193)
	at crux.http_server$handler.invokeStatic(http_server.clj:320)
	at crux.http_server$handler.invoke(http_server.clj:278)
	at crux.http_server$__GT_server$fn__9501.invoke(http_server.clj:415)
	at clojure.core$some_fn$sp3__8668.invoke(core.clj:7461)
	at ring.middleware.params$wrap_params$fn__9289.invoke(params.clj:67)
	at ring.middleware.resource$wrap_resource_prefer_resources$fn__9325.invoke(resource.clj:25)
	at crux.http_server$wrap_exception_handling$fn__9354.invoke(http_server.clj:75)
	at ring.adapter.jetty$proxy_handler$fn__9079.invoke(jetty.clj:27)
	at ring.adapter.jetty.proxy$org.eclipse.jetty.server.handler.AbstractHandler$ff19274a.handle(Unknown Source)
	at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:127)
	at org.eclipse.jetty.server.Server.handle(Server.java:500)
	at org.eclipse.jetty.server.HttpChannel.lambda$handle$1(HttpChannel.java:386)
	at org.eclipse.jetty.server.HttpChannel.dispatch(HttpChannel.java:562)
	at org.eclipse.jetty.server.HttpChannel.handle(HttpChannel.java:378)
	at org.eclipse.jetty.server.HttpConnection.onFillable(HttpConnection.java:270)
	at org.eclipse.jetty.io.AbstractConnection$ReadCallback.succeeded(AbstractConnection.java:311)
	at org.eclipse.jetty.io.FillInterest.fillable(FillInterest.java:103)
	at org.eclipse.jetty.io.ChannelEndPoint$2.run(ChannelEndPoint.java:117)
	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.runTask(EatWhatYouKill.java:336)
	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.doProduce(EatWhatYouKill.java:313)
	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.tryProduce(EatWhatYouKill.java:171)
	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.produce(EatWhatYouKill.java:135)
	at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:806)
	at org.eclipse.jetty.util.thread.QueuedThreadPool$Runner.run(QueuedThreadPool.java:938)
	at java.base/java.lang.Thread.run(Thread.java:834)
```

Project.clj
```clojure
(defproject test-crux "0.1.0-SNAPSHOT"
  :description "CRUX db"
  :dependencies [[org.clojure/clojure "1.10.1"]
                 [org.slf4j/slf4j-api "1.7.29"]
                 [ch.qos.logback/logback-classic "1.2.3"]
                 [juxt/crux-core "20.09-1.12.1-beta"]
                 [juxt/crux-kafka "20.09-1.12.1-beta"]
                 [juxt/crux-rocksdb "20.09-1.12.1-beta"]
                 [juxt/crux-http-server "20.09-1.12.1-alpha"]
                 [environ "1.2.0"]
                 ]
  :main ^:skip-aot test-crux.core
  :target-path "target/%s"
  :profiles {:uberjar {:aot :all}})
```

Main file:
```clojure
(ns test-crux.core
  (:require [crux.api :as crux]
            [clojure.java.io :as io]
            [environ.core :refer [env]])
  (:gen-class))
```

Configuration:
```clojure
(defn start-node
  [kafka-host kafka-port kafka-properties-file storage-dir http-server-port]
  (let [config {:rocksdb-config          {:crux/module 'crux.rocksdb/->kv-store
                                          :db-dir      (str (io/file storage-dir "db"))}
                :kafka-config            {:crux/module       'crux.kafka/->kafka-config
                                          :bootstrap-servers (str kafka-host ":" kafka-port)
                                          :properties-file   kafka-properties-file
                                          }
                :crux/index-store        {:kv-store :rocksdb-config}
                :crux/tx-log             {:crux/module  'crux.kafka/->tx-log
                                          :kafka-config :kafka-config
                                          :tx-topic-opts {:topic-name "crux-transaction-log-json"}}
                :crux/document-store     {:crux/module          'crux.kafka/->document-store
                                          :kafka-config         :kafka-config
                                          :local-document-store {:kv-store :rocksdb-config}
                                          :doc-topic-opts {:topic-name "crux-docs-json"}}
                :crux.http-server/server {:port http-server-port}}]
    (crux/start-node config)))
```

Properties only this addition (in comparison to default Nippy config):
```properties
key.serializer=crux.kafka.json.JsonSerializer
key.deserializer=crux.kafka.json.JsonDeserializer
value.serializer=crux.kafka.json.JsonSerializer
value.deserializer=crux.kafka.json.JsonDeserializer
```

**Comment by lavsurgut** [comment]
So sending any document tx via default Nippy SerDe works ok.

With JSON though:

ERROR crux.http-server - Exception while handling request: {:ssl-client-cert nil, :protocol "HTTP/1.1", :remote-addr "10.0.11.28", :headers {"user-agent" "Go-http-client/1.1", "host" "crux:3000", "accept-encoding" "gzip", "content-length" "1131", "content-type" "application/edn"}, :server-port 3000, :content-length 1131, :content-type "application/edn", :character-encoding nil, :uri "/tx-log", :server-name "crux", :query-string nil, :body #object[org.eclipse.jetty.server.HttpInputOverHTTP 0x63b03bf0 "HttpInputOverHTTP@63b03bf0[c=1131,q=0,[0]=null,s=EOF]"], :scheme :http, :request-method :post}
com.fasterxml.jackson.core.JsonGenerationException: Cannot JSON encode object of class: class crux.codec.Id: d827db3da2a4b618261eaba80cdf5f1f584609ee
	at cheshire.generate$generate.invokeStatic(generate.clj:152)
	at cheshire.generate$generate.invoke(generate.clj:116)
	at cheshire.core$generate_string.invokeStatic(core.clj:74)
	at cheshire.core$generate_string.invoke(core.clj:49)
	at cheshire.core$generate_string.invokeStatic(core.clj:55)
	at cheshire.core$generate_string.invoke(core.clj:49)
	at clojure.lang.Var.invoke(Var.java:384)
	at crux.kafka.json.JsonSerializer.serialize(JsonSerializer.java:53)
	at org.apache.kafka.common.serialization.Serializer.serialize(Serializer.java:62)
	at org.apache.kafka.clients.producer.KafkaProducer.doSend(KafkaProducer.java:886)
	at org.apache.kafka.clients.producer.KafkaProducer.send(KafkaProducer.java:856)
	at org.apache.kafka.clients.producer.KafkaProducer.send(KafkaProducer.java:743)
	at crux.kafka$submit_docs.invokeStatic(kafka.clj:231)
	at crux.kafka$submit_docs.invoke(kafka.clj:228)
	at crux.kafka.KafkaDocumentStore.submit_docs(kafka.clj:245)
	at crux.node.CruxNode.submitTx(node.clj:125)
	at crux.http_server$transact.invokeStatic(http_server.clj:195)
	at crux.http_server$transact.invoke(http_server.clj:193)
	at crux.http_server$handler.invokeStatic(http_server.clj:320)
	at crux.http_server$handler.invoke(http_server.clj:278)
	at crux.http_server$__GT_server$fn__9501.invoke(http_server.clj:415)
	at clojure.core$some_fn$sp3__8668.invoke(core.clj:7461)
	at ring.middleware.params$wrap_params$fn__9289.invoke(params.clj:67)
	at ring.middleware.resource$wrap_resource_prefer_resources$fn__9325.invoke(resource.clj:25)
	at crux.http_server$wrap_exception_handling$fn__9354.invoke(http_server.clj:75)
	at ring.adapter.jetty$proxy_handler$fn__9079.invoke(jetty.clj:27)
	at ring.adapter.jetty.proxy$org.eclipse.jetty.server.handler.AbstractHandler$ff19274a.handle(Unknown Source)
	at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:127)
	at org.eclipse.jetty.server.Server.handle(Server.java:500)
	at org.eclipse.jetty.server.HttpChannel.lambda$handle$1(HttpChannel.java:386)
	at org.eclipse.jetty.server.HttpChannel.dispatch(HttpChannel.java:562)
	at org.eclipse.jetty.server.HttpChannel.handle(HttpChannel.java:378)
	at org.eclipse.jetty.server.HttpConnection.onFillable(HttpConnection.java:270)
	at org.eclipse.jetty.io.AbstractConnection$ReadCallback.succeeded(AbstractConnection.java:311)
	at org.eclipse.jetty.io.FillInterest.fillable(FillInterest.java:103)
	at org.eclipse.jetty.io.ChannelEndPoint$2.run(ChannelEndPoint.java:117)
	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.runTask(EatWhatYouKill.java:336)
	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.doProduce(EatWhatYouKill.java:313)
	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.tryProduce(EatWhatYouKill.java:171)
	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.produce(EatWhatYouKill.java:135)
	at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:806)
	at org.eclipse.jetty.util.thread.QueuedThreadPool$Runner.run(QueuedThreadPool.java:938)
	at java.base/java.lang.Thread.run(Thread.java:834)

**Comment by lavsurgut** [comment]
Setting properties to Transit:

```properties
key.serializer=crux.kafka.transit.TransitSerializer
key.deserializer=crux.kafka.transit.TransitDeserializer
value.serializer=crux.kafka.transit.TransitSerializer
value.deserializer=crux.kafka.transit.TransitDeserializer
```

Produces a different error:

```clojure
(crux/submit-tx node [[:crux.tx/put {:crux.db/id :ivan :name "Ivan" :homeworld "Earth"}]])

;; Execution error (NullPointerException) at com.cognitect.transit.impl.AbstractEmitter/marshalTop (AbstractEmitter.java:203).
null
```

**Comment by lavsurgut** [comment]
Same errors happen also via clojure api calls

**Comment by jarohen** [comment]
(moved the first comment up in to the main issue description)

**Comment by jarohen** [comment]
Merged #1195 to fix the Transit serdes, thanks @danmason :)

**Comment by refset** [comment]
Hi @lavsurgut, as per our Zulip DM exchange, we are going to remove this JSON serdes option for the time being due to bandwidth constraints & the additional design work necessary before we can "fix" it. We also discussed how Transit's `:json-verbose` mode might be sufficient for your immediate needs - hopefully that works out okay :)

We will push a commit shortly to remove the relevant files and then close this issue. If anyone determines that they really need JSON serdes in future then let's use https://github.com/juxt/crux/issues/151 or open a new issue to discuss the requirement. 

Thanks again for bringing this to our attention :pray:
